hal/src/sercom/v2/dma.rs

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    #[inline]
+    fn incrementing(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    fn buffer_len(&self) -> usize {
+        1
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    Self: Buffer<Beat = C::Word>,
+    C: uart::ValidConfig,
+    D: Receive,
+{
+    /// Transform an [`Uart`] into a DMA [`Transfer`]) and
+    /// start receiving into the provided buffer.
+    #[inline]
+    pub fn receive_with_dma<Ch, B, W>(
+        self,
+        buf: B,
+        mut channel: Ch,
+        waker: W,
+    ) -> Transfer<Channel<Ch::Id, Busy>, BufferPair<Self, B>, W>
+    where
+        Ch: AnyChannel<Status = Ready>,
+        B: Buffer<Beat = C::Word> + 'static,
+        W: FnOnce(CallbackStatus) + 'static,
+    {
+        channel
+            .as_mut()
+            .enable_interrupts(InterruptFlags::new().with_tcmpl(true));
+
+        #[cfg(feature = "min-samd51g")]
+        let trigger_action = TriggerAction::BURST;
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        let trigger_action = TriggerAction::BEAT;
+
+        // SAFETY: We use new_unchecked to avoid having to pass a 'static self as the
+        // destination buffer. This is safe as long as we guarantee the source buffer is
+        // static.
+        unsafe { dmac::Transfer::new_unchecked(channel, self, buf, false) }
+            .with_waker(waker)
+            .begin(C::Sercom::DMA_RX_TRIGGER, trigger_action)
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    Self: Buffer<Beat = C::Word>,
+    C: uart::ValidConfig,
+    D: Transmit,
+{
+    /// Transform an [`Uart`] into a DMA [`Transfer`]) and
+    /// start sending the provided buffer.
+    #[inline]
+    pub fn send_with_dma<Ch, B, W>(
+        self,
+        buf: B,
+        mut channel: Ch,
+        waker: W,
+    ) -> Transfer<Channel<Ch::Id, Busy>, BufferPair<B, Self>, W>
+    where
+        Ch: AnyChannel<Status = Ready>,
+        B: Buffer<Beat = C::Word> + 'static,
+        W: FnOnce(CallbackStatus) + 'static,
+    {
+        channel
+            .as_mut()
+            .enable_interrupts(InterruptFlags::new().with_tcmpl(true));
+
+        #[cfg(feature = "min-samd51g")]
+        let trigger_action = TriggerAction::BURST;
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        let trigger_action = TriggerAction::BEAT;
+
+        // SAFETY: We use new_unchecked to avoid having to pass a 'static self as the
+        // destination buffer. This is safe as long as we guarantee the source buffer is
+        // static.
+        unsafe { dmac::Transfer::new_unchecked(channel, buf, self, false) }
+            .with_waker(waker)
+            .begin(C::Sercom::DMA_TX_TRIGGER, trigger_action)
+    }
+}
+
+//=============================================================================
+// SPI DMA transfers
+//=============================================================================
+
+#[cfg(feature = "min-samd51g")]
+#[doc(hidden)]
+pub trait CharSize: spi::Length {}
```

Comment:
This isn't going to work. I think your DMA approach here will only work for SAMD5x chips when `Length` is `U4` or less. For those lengths, each write to the `DATA` register is essentially atomic. It represents a complete transaction. But for lengths greater than `U4`, the SPI peripheral keeps track of the state of the transfer in its `LENGTH` register, so your concept of `buffer_len = 1` doesn't match.

In general, I think it might make more sense to define respective implementations in the `thumbv*` modules and re-export them here. Otherwise, you'll end up with a ton of `cfg`s.

Reply:
I think adding an `AtomicLength` trait, which limits acceptable `Lengths` to U1 through U4, fixes that.

Reply:
I'm reworking the `spi` module to add `Capability` right now, so why don't we hold off on DMA for `spi` for the moment. I think your proposed solution works though.